platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/recovery/internal/EventStreamSingleFileRepairIterator.java

```diff
@@ -0,0 +1,196 @@
+
+    /**
+     * The logger to record information and errors.
+     */
+    private final Logger logger = LogManager.getLogger(EventStreamSingleFileRepairIterator.class);
+    /**
+     * The sequence of SelfSerializables that may need a final hash added.
+     */
+    private final IOIterator<SelfSerializable> eventStreamIterator;
+    /**
+     * The running hash calculator for computing the final hash.
+     */
+    private final RunningHashCalculatorForStream<DetailedConsensusEvent> runningHashCalculator;
+    /**
+     * The next event to return in the event stream.
+     */
+    private SelfSerializable next = null;
+    /**
+     * The most recent event returned in the event stream.
+     */
+    private SelfSerializable prev = null;
+    /**
+     * indicates the number of hashes we've seen through the event stream iterator.
+     */
+    private int hashCount = 0;
+    /**
+     * A boolean to indicate if we provided a new final hash for the event stream iterator.
+     */
+    private boolean finalHashAdded = false;
+    /**
+     * The count of events in the event stream.
+     */
+    private int eventCount = 0;
+
+    /**
+     * Constructs a repairing iterator for the given event stream iterator.
+     *
+     * @param eventStream The event stream iterator that may be missing a final hash.
+     */
+    public EventStreamSingleFileRepairIterator(final IOIterator<SelfSerializable> eventStream) {
+        this.eventStreamIterator = Objects.requireNonNull(eventStream, "The eventStream iterator must not be null.");
+        this.runningHashCalculator = new RunningHashCalculatorForStream<>();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean hasNext() {
+        return next != null || findNextWithRepair();
+    }
+
+    /**
+     * Sets the next SelfSerializable from the event stream iterator, increments the count of hashes or events, and
+     * updates the running hash.  If the event stream iterator does not have a next element or throws an exception, the
+     * next variable is set to null.
+     */
+    private void setNextAndUpdateRunningHash() {
+        try {
+            if (eventStreamIterator.hasNext()) {
+                next = eventStreamIterator.next();
+                if (next instanceof Hash hash) {
+                    hashCount++;
+                    if (hashCount == 1) {
+                        // seed running hash calculator with first hash.
+                        runningHashCalculator.setRunningHash(hash);
+                    }
+                } else if (next instanceof DetailedConsensusEvent dce) {
+                    eventCount++;
+                    // update running hash calculator with event.
+                    runningHashCalculator.addObject(dce);
+                } else {
+                    logger.warn(
+                            "Unexpected event in repair iterator, class: {}",
+                            next.getClass().getCanonicalName());
```

Comment:
No log marker... but probably ok to just throw here.

Reply:
now throwing an IllegalStateException